### PR TITLE
Adds option to log to file

### PR DESCRIPTION
**Pull request by breakreturn**
Adds a cmake option ENABLE_LOG_TO_FILE
that makes the proxy log to a file rather
than stdout/stderr. This also changes the
code to use e.g. g_message() instead of
g_print().

Signed-off-by: Joakim Gross <joakim.gross@pelagicore.com>

### CODE REVIEW

**Review by frznlogic — APPROVED** [review]

File: src/proxy.c [hunk 1]
```diff
@@ -149,7 +145,7 @@ DBusHandlerResult filter_cb (DBusConnection *conn,
 		              dbus_message_get_path      (msg),
 		              dbus_message_get_member    (msg)))
 	{
-		g_print ("Accepted call to '%s' from "
+		g_message("Accepted call to '%s' from "
 		                "client to '%s' on '%s'.\n",
 		        dbus_message_get_member   (msg),
```
Comment (kursatkobya) [inline src/proxy.c]: indentation seems broken

File: src/proxy.c [hunk 2]
```diff
@@ -368,17 +363,17 @@ DBusHandlerResult master_filter_cb (DBusConnection *conn,
...
 		if (dest != NULL &&
 		    is_conn_known_eavesdropper(dest))
 		{
 			if (verbose)
 			{
-				g_print("New connection's unique name ('%s')"
-					" was previously known as an eavesdropper."
-					" Removed old entry...\n", dest);
+				g_message("New connection's unique name ('%s')"
+                          " was previously known as an eavesdropper."
```
Comment (kursatkobya) [inline src/proxy.c]: indentation

File: src/proxy.c [hunk 3]
```diff
@@ -401,24 +396,24 @@ DBusHandlerResult master_filter_cb (DBusConnection *conn,
 	} else if (is_conn_known_eavesdropper (dbus_bus_get_unique_name(conn)))
 	{
 		if (verbose) {
-			g_print ("'%s' is an eavesdropping connection, let it go...\n",
-				 dbus_bus_get_unique_name(conn));
+			g_message("'%s' is an eavesdropping connection, let it go...\n",
+				      dbus_bus_get_unique_name(conn));
```
Comment (kursatkobya) [inline src/proxy.c]: indentation

File: src/proxy.c [hunk 4]
```diff
@@ -467,9 +462,9 @@ gboolean is_incoming_eavesdropping (DBusMessage *msg)
 			is_eavesdropping = TRUE;
 			if (verbose)
 			{
-				g_print ("'%s' AddMatch-args: \"%s\"\n",
-					 dbus_message_get_sender(msg),
-					 msg_arguments);
+				g_message("'%s' AddMatch-args: \"%s\"\n",
```
Comment (kursatkobya) [inline src/proxy.c]: indentation

File: src/proxy.c [hunk 5]
```diff
@@ -689,10 +683,55 @@ int parse_json_from_stdin (const char *section) {
 
 void print_usage()
 {
-    g_print("dbus-proxy, version %s\n", PACKAGE_VERSION);
-    g_print("Usage: dbus-proxy address session|system\n"
-            "waits for JSON conf at stdin.\n");
+    g_message("dbus-proxy, version %s\n", PACKAGE_VERSION);
+    g_message("Usage: dbus-proxy address session|system\n"
+              "waits for JSON conf at stdin.\n");
```
Comment (tobsan) [inline src/proxy.c]: Do we really want to use the log handler for this output? Shouldn't this always go to stdout?

File: src/proxy.c [hunk 6]
```diff
@@ -689,10 +683,55 @@ int parse_json_from_stdin (const char *section) {
...
+FILE *log_file;
+
+gboolean log_file_is_open() {
+    return log_file ? TRUE : FALSE;
+}
+
+gboolean open_log_file() {
+    log_file = fopen("/tmp/dbus-proxy-log", "a");
+
+    if (log_file == NULL) {
```
Comment (kursatkobya) [inline src/proxy.c]: `NULL == log_file` would be better

File: src/proxy.c [hunk 7]
```diff
@@ -737,10 +776,20 @@ int main(int argc, char *argv[]) {
 		exit(1);
 	}
 
+	/* Parse JSON */
+	if (parse_json_from_stdin (argv[2]) == 1){
```
Comment (kursatkobya) [inline src/proxy.c]: `1 == ...` would be better

**Review by kursatkobya — APPROVED** [review]
Approved with minor comments

File: src/proxy.c [hunk 8]
```diff
@@ -717,15 +756,15 @@ int main(int argc, char *argv[]) {
...
 	}
 
-	/* Parse JSON */
-	if (parse_json_from_stdin (argv[2]) == 1){
-		g_print ("Something wrong with JSON file. Exiting...\n");
-		exit (1);
-	}
+#ifdef LOG_TO_FILE
+    /* Set log handler for e.g. g_message(), g_warning() etc. */
+    open_log_file();
```
Comment (tobsan) [inline src/proxy.c]: Perhaps you should check the return value of this.

File: src/proxy.c [hunk 9]
```diff
@@ -689,10 +683,55 @@ int parse_json_from_stdin (const char *section) {
...
+    log_file = fopen("/tmp/dbus-proxy-log", "a");
+
+    if (log_file == NULL) {
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+gboolean close_log_file() {
```
Comment (tobsan) [inline src/proxy.c:715]: Perhaps it would make sense to check if the `log_file` is open before you close it?

**Review by tobsan — CHANGES_REQUESTED** [review]

File: README.md [hunk 10]
```diff
@@ -26,6 +26,11 @@ $ make
 $ make install
 ```
 
+The following cmake options are available:
+
+* `ENABLE_LOG_TO_FILE` - makes `dbus-proxy` log to "/tmp/dbus-proxy-log"
```
Comment (tobsan) [inline README.md]: Why not `/tmp/dbus-proxy.log`?

**Review by tobsan — APPROVED** [review]

**Pull request merged**